Scripts/Runtime/Data/VirtualData/Jobs/VDJobUpdater.cs

```diff
@@ -0,0 +1,121 @@
+    /// <see cref="Complete"/> if the work is done.
+    /// </summary>
+    /// <typeparam name="TKey">The type of key to use for lookup of the instance</typeparam>
+    /// <typeparam name="TInstance">The type of instance</typeparam>
+    [BurstCompatible]
+    public struct VDJobUpdater<TKey, TInstance>
+        where TKey : struct, IEquatable<TKey>
+        where TInstance : struct, ILookupData<TKey>
+    {
+        private const int DEFAULT_LANE_INDEX = -1;
+
+        [ReadOnly] private readonly UnsafeTypedStream<TInstance>.Writer m_ContinueWriter;
+        [ReadOnly] private readonly NativeArray<TInstance> m_Iteration;
+
+        private UnsafeTypedStream<TInstance>.LaneWriter m_ContinueLaneWriter;
+
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+        private enum UpdaterState
+        {
+            Uninitialized,
+            Ready,
+            Modifying
+        }
+
+        private UpdaterState m_State;
+#endif
+
+        public int LaneIndex
+        {
+            get;
+            private set;
+        }
+
+        internal VDJobUpdater(UnsafeTypedStream<TInstance>.Writer continueWriter,
+                              NativeArray<TInstance> iteration)
+        {
+            m_ContinueWriter = continueWriter;
+            m_Iteration = iteration;
+
+            m_ContinueLaneWriter = default;
+            LaneIndex = DEFAULT_LANE_INDEX;
+
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+            m_State = UpdaterState.Uninitialized;
+#endif
+        }
+        
+        /// <summary>
+        /// Initializes the struct based on the thread it's being used on.
+        /// This must be called before doing anything else with the struct.
+        /// </summary>
+        /// <param name="nativeThreadIndex">The native thread index</param>
+        public void InitForThread(int nativeThreadIndex)
+        {
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+            Debug.Assert(m_State == UpdaterState.Uninitialized);
+            m_State = UpdaterState.Ready;
+#endif
+
+            LaneIndex = ParallelAccessUtil.CollectionIndexForThread(nativeThreadIndex);
+            m_ContinueLaneWriter = m_ContinueWriter.AsLaneWriter(LaneIndex);
+        }
+        
+        /// <summary>
+        /// Gets a <typeparamref name="TInstance"/> at the specified index.
+        /// </summary>
+        /// <param name="index">The index to the backing array</param>
+        public TInstance this[int index]
+        {
+            get
+            {
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+                Debug.Assert(m_State == UpdaterState.Ready);
+                m_State = UpdaterState.Modifying;
+#endif
+
+                return m_Iteration[index];
+            }
+        }
+        
+        /// <summary>
+        /// Signals that this instance should be updated again next frame.
+        /// </summary>
+        /// <param name="instance">The instance to continue</param>
+        public void Continue(TInstance instance)
+        {
+            Continue(ref instance);
```

Comment:
I'm not sure how Burst factors into this, but a lot of the internal Unity code is using `ref` when passing around structs.

Additionally, because of our layers of abstraction, `Continue` then calls `Write` which then calls `UnsafeUtility.CopyStructureToPtr`. So instead of multiple copies, we can just pass the same pointer all the way down which is what it wants at the end anyway. 

Agreed that in C# if the struct is small, then it's better to just copy, but multiple copies? Seems like that might get into the tiny allocate/deallocate cost on the stack mentioned here: https://stackoverflow.com/questions/46291828/when-is-it-more-efficient-to-pass-structs-by-value-and-when-by-ref-in-c

Reply:
That's fair. You lose data locality passing refs but unless you're storing that ref in a big collection and iterating later it's not a big deal.

Reply:
Seems right to me.